<h1>Activity 1 - PALMER PENGUINS DATA ANALYITCS PROJECT</h1>
<br>
<h3>Analyst: Ron Jerick Gamba</h3>

In [1]:
import numpy as np 
import pandas as pd

<h1>Converting csv into a Data Frame</h1>
<br>
<h3>Requirements</h3>
<ul>
    <li>Data Frame name</li>
    <li>CSV File path</li>
</ul>

In [2]:
penguins = pd.read_csv('datasets//penguins.csv')
penguins

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0       Adelie  Torgersen            39.1           18.7              181.0   
1       Adelie  Torgersen            39.5           17.4              186.0   
2       Adelie  Torgersen            40.3           18.0              195.0   
3       Adelie  Torgersen             NaN            NaN                NaN   
4       Adelie  Torgersen            36.7           19.3              193.0   
..         ...        ...             ...            ...                ...   
339  Chinstrap      Dream            55.8           19.8              207.0   
340  Chinstrap      Dream            43.5           18.1              202.0   
341  Chinstrap      Dream            49.6           18.2              193.0   
342  Chinstrap      Dream            50.8           19.0              210.0   
343  Chinstrap      Dream            50.2           18.7              198.0   

     body_mass_g     sex  year  
0         3750.0    male  2007  
1         3800.0  female  2007  
2         3250.0  female  2007  
3            NaN     NaN  2007  
4         3450.0  female  2007  
..           ...     ...   ...  
339       4000.0    male  2009  
340       3400.0  female  2009  
341       3775.0    male  2009  
342       4100.0    male  2009  
343       3775.0  female  2009  

[344 rows x 8 columns]

<h1>Describe Data</h1>

In [3]:
penguins.describe()

bill_length_mm  bill_depth_mm  flipper_length_mm  body_mass_g  \
count      342.000000     342.000000         342.000000   342.000000   
mean        43.921930      17.151170         200.915205  4201.754386   
std          5.459584       1.974793          14.061714   801.954536   
min         32.100000      13.100000         172.000000  2700.000000   
25%         39.225000      15.600000         190.000000  3550.000000   
50%         44.450000      17.300000         197.000000  4050.000000   
75%         48.500000      18.700000         213.000000  4750.000000   
max         59.600000      21.500000         231.000000  6300.000000   

              year  
count   344.000000  
mean   2008.029070  
std       0.818356  
min    2007.000000  
25%    2007.000000  
50%    2008.000000  
75%    2009.000000  
max    2009.000000

In [4]:
penguins.describe().T

count         mean         std     min       25%      50%  \
bill_length_mm     342.0    43.921930    5.459584    32.1    39.225    44.45   
bill_depth_mm      342.0    17.151170    1.974793    13.1    15.600    17.30   
flipper_length_mm  342.0   200.915205   14.061714   172.0   190.000   197.00   
body_mass_g        342.0  4201.754386  801.954536  2700.0  3550.000  4050.00   
year               344.0  2008.029070    0.818356  2007.0  2007.000  2008.00   

                      75%     max  
bill_length_mm       48.5    59.6  
bill_depth_mm        18.7    21.5  
flipper_length_mm   213.0   231.0  
body_mass_g        4750.0  6300.0  
year               2009.0  2009.0

<h1>Checking Null Values</h1>

In [5]:
penguins.isnull().sum()

species               0
island                0
bill_length_mm        2
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11
year                  0
dtype: int64

<h1>Checking Values Counts</h1>

In [6]:
penguins['species']

0         Adelie
1         Adelie
2         Adelie
3         Adelie
4         Adelie
         ...    
339    Chinstrap
340    Chinstrap
341    Chinstrap
342    Chinstrap
343    Chinstrap
Name: species, Length: 344, dtype: object

In [7]:
penguins['species'].value_counts()

species
Adelie       152
Gentoo       124
Chinstrap     68
Name: count, dtype: int64

<h1>Filtering Data</h1>

In [8]:
adelie_DF = penguins[penguins['species'] == "Adelie"]
adelie_DF

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0    Adelie  Torgersen            39.1           18.7              181.0   
1    Adelie  Torgersen            39.5           17.4              186.0   
2    Adelie  Torgersen            40.3           18.0              195.0   
3    Adelie  Torgersen             NaN            NaN                NaN   
4    Adelie  Torgersen            36.7           19.3              193.0   
..      ...        ...             ...            ...                ...   
147  Adelie      Dream            36.6           18.4              184.0   
148  Adelie      Dream            36.0           17.8              195.0   
149  Adelie      Dream            37.8           18.1              193.0   
150  Adelie      Dream            36.0           17.1              187.0   
151  Adelie      Dream            41.5           18.5              201.0   

     body_mass_g     sex  year  
0         3750.0    male  2007  
1         3800.0  female  2007  
2         3250.0  female  2007  
3            NaN     NaN  2007  
4         3450.0  female  2007  
..           ...     ...   ...  
147       3475.0  female  2009  
148       3450.0  female  2009  
149       3750.0    male  2009  
150       3700.0  female  2009  
151       4000.0    male  2009  

[152 rows x 8 columns]

In [9]:
gentoo_DF = penguins[penguins['species'] == "Gentoo"]
gentoo_DF

species  island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
152  Gentoo  Biscoe            46.1           13.2              211.0   
153  Gentoo  Biscoe            50.0           16.3              230.0   
154  Gentoo  Biscoe            48.7           14.1              210.0   
155  Gentoo  Biscoe            50.0           15.2              218.0   
156  Gentoo  Biscoe            47.6           14.5              215.0   
..      ...     ...             ...            ...                ...   
271  Gentoo  Biscoe             NaN            NaN                NaN   
272  Gentoo  Biscoe            46.8           14.3              215.0   
273  Gentoo  Biscoe            50.4           15.7              222.0   
274  Gentoo  Biscoe            45.2           14.8              212.0   
275  Gentoo  Biscoe            49.9           16.1              213.0   

     body_mass_g     sex  year  
152       4500.0  female  2007  
153       5700.0    male  2007  
154       4450.0  female  2007  
155       5700.0    male  2007  
156       5400.0    male  2007  
..           ...     ...   ...  
271          NaN     NaN  2009  
272       4850.0  female  2009  
273       5750.0    male  2009  
274       5200.0  female  2009  
275       5400.0    male  2009  

[124 rows x 8 columns]

In [10]:
chinstrap_DF = penguins[penguins['species'] == "Chinstrap"]
chinstrap_DF

species island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
276  Chinstrap  Dream            46.5           17.9              192.0   
277  Chinstrap  Dream            50.0           19.5              196.0   
278  Chinstrap  Dream            51.3           19.2              193.0   
279  Chinstrap  Dream            45.4           18.7              188.0   
280  Chinstrap  Dream            52.7           19.8              197.0   
..         ...    ...             ...            ...                ...   
339  Chinstrap  Dream            55.8           19.8              207.0   
340  Chinstrap  Dream            43.5           18.1              202.0   
341  Chinstrap  Dream            49.6           18.2              193.0   
342  Chinstrap  Dream            50.8           19.0              210.0   
343  Chinstrap  Dream            50.2           18.7              198.0   

     body_mass_g     sex  year  
276       3500.0  female  2007  
277       3900.0    male  2007  
278       3650.0    male  2007  
279       3525.0  female  2007  
280       3725.0    male  2007  
..           ...     ...   ...  
339       4000.0    male  2009  
340       3400.0  female  2009  
341       3775.0    male  2009  
342       4100.0    male  2009  
343       3775.0  female  2009  

[68 rows x 8 columns]

<h1>Questions</h1>
<br> 
<ul>
    <li>
        How many of these Male Chinstrap species weight is abouve 4000 grams
    </li>
</ul>

In [11]:
heavy_chinstrap = penguins[(penguins['species'] == 'Chinstrap') & (penguins['body_mass_g'] >= 4000)]
heavy_chinstrap

species island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
284  Chinstrap  Dream            46.0           18.9              195.0   
289  Chinstrap  Dream            52.0           18.1              201.0   
291  Chinstrap  Dream            50.5           19.6              201.0   
295  Chinstrap  Dream            49.2           18.2              195.0   
301  Chinstrap  Dream            52.0           19.0              197.0   
305  Chinstrap  Dream            52.8           20.0              205.0   
307  Chinstrap  Dream            54.2           20.8              201.0   
309  Chinstrap  Dream            51.0           18.8              203.0   
313  Chinstrap  Dream            52.0           20.7              210.0   
315  Chinstrap  Dream            53.5           19.9              205.0   
321  Chinstrap  Dream            50.8           18.5              201.0   
323  Chinstrap  Dream            49.0           19.6              212.0   
329  Chinstrap  Dream            50.7           19.7              203.0   
333  Chinstrap  Dream            49.3           19.9              203.0   
339  Chinstrap  Dream            55.8           19.8              207.0   
342  Chinstrap  Dream            50.8           19.0              210.0   

     body_mass_g     sex  year  
284       4150.0  female  2007  
289       4050.0    male  2007  
291       4050.0    male  2007  
295       4400.0    male  2007  
301       4150.0    male  2007  
305       4550.0    male  2008  
307       4300.0    male  2008  
309       4100.0    male  2008  
313       4800.0    male  2008  
315       4500.0    male  2008  
321       4450.0    male  2009  
323       4300.0    male  2009  
329       4050.0    male  2009  
333       4050.0    male  2009  
339       4000.0    male  2009  
342       4100.0    male  2009

<h1>Fillin up Null Values</h1>

In [12]:
penguins.isnull().sum()

species               0
island                0
bill_length_mm        2
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11
year                  0
dtype: int64

In [13]:
average_bill_len = penguins['bill_length_mm'].mean()
penguins['bill_length_mm'] = penguins['bill_length_mm'].fillna(average_bill_len)
penguins.isnull().sum()

species               0
island                0
bill_length_mm        0
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11
year                  0
dtype: int64

In [14]:
average_bill_depth = penguins['bill_depth_mm'].mean()
penguins['bill_depth_mm'] = penguins['bill_depth_mm'].fillna(average_bill_depth)
penguins.isnull().sum()

species               0
island                0
bill_length_mm        0
bill_depth_mm         0
flipper_length_mm     2
body_mass_g           2
sex                  11
year                  0
dtype: int64

In [15]:
average_flipper_ave = penguins['flipper_length_mm'].mean()
penguins['flipper_length_mm'] = penguins['flipper_length_mm'].fillna(average_flipper_ave)
penguins.isnull().sum()

species               0
island                0
bill_length_mm        0
bill_depth_mm         0
flipper_length_mm     0
body_mass_g           2
sex                  11
year                  0
dtype: int64

In [16]:
average_body_mass = penguins['body_mass_g'].mean()

In [17]:
penguins['body_mass_g'] = penguins['body_mass_g'].fillna(average_body_mass)
penguins.isnull().sum()

species               0
island                0
bill_length_mm        0
bill_depth_mm         0
flipper_length_mm     0
body_mass_g           0
sex                  11
year                  0
dtype: int64

In [18]:
frequency_sex = penguins['sex'].mode()
penguins['sex'] = penguins['sex'].fillna(frequency_sex[0])
penguins.isnull().sum()

species              0
island               0
bill_length_mm       0
bill_depth_mm        0
flipper_length_mm    0
body_mass_g          0
sex                  0
year                 0
dtype: int64

In [19]:
penguins.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   species            344 non-null    object 
 1   island             344 non-null    object 
 2   bill_length_mm     344 non-null    float64
 3   bill_depth_mm      344 non-null    float64
 4   flipper_length_mm  344 non-null    float64
 5   body_mass_g        344 non-null    float64
 6   sex                344 non-null    object 
 7   year               344 non-null    int64  
dtypes: float64(4), int64(1), object(3)
memory usage: 21.6+ KB


In [20]:
penguins['species'] = penguins['species'].astype('category')
penguins['island'] = penguins['island'].astype('category') 
penguins['sex'] = penguins['sex'].astype('category')
penguins.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype   
---  ------             --------------  -----   
 0   species            344 non-null    category
 1   island             344 non-null    category
 2   bill_length_mm     344 non-null    float64 
 3   bill_depth_mm      344 non-null    float64 
 4   flipper_length_mm  344 non-null    float64 
 5   body_mass_g        344 non-null    float64 
 6   sex                344 non-null    category
 7   year               344 non-null    int64   
dtypes: category(3), float64(4), int64(1)
memory usage: 15.0 KB


In [21]:
penguins['mass_kg'] = np.round(penguins['body_mass_g'] / 1000, 2)

<h1>Exploratory Data Analysis</h1>
<Hr>
<ol>
  <li>What is the average bill depth of each penguin species?</li>
  <li>Which sex in the Chinstrap species has the longest average flipper length?</li>
  <li>In what year was the shortest flipper length recorded in an Adelie penguin?</li>
  <li>Which island has the heaviest average body mass of penguins?</li>
  <li>What is the average bill length of Gentoo penguins per year?</li>
  <li>How many Chinstrap penguins were recorded in 2008, and which one had the longest bill?</li>
  <li>What is the most common penguin species found on Dream Island?</li>
  <li>What is the smallest flipper length recorded on Biscoe Island?</li>
  <li>What is the largest bill depth recorded on Torgersen Island?</li>
  <li>Which year had the highest average body mass of penguins on Dream Island?</li>
  <li>How many Chinstrap penguins are found on Dream Island?</li>
  <li>What is the average body mass of male penguins on Biscoe Island?</li>
  <li>What is the longest bill length recorded on Dream Island?</li>
</ol>


<h1>1.What is the average bill depth of each penguin species?</h1>

In [22]:
avg_bill_depth_adelie = penguins[penguins['species'] == 'Adelie']['bill_depth_mm'].mean()
avg_bill_depth_chinstrap = penguins[penguins['species'] == 'Chinstrap']['bill_depth_mm'].mean()
avg_bill_depth_gentoo = penguins[penguins['species'] == 'Gentoo']['bill_depth_mm'].mean()

print("Adelie:", avg_bill_depth_adelie)
print("Chinstrap:", avg_bill_depth_chinstrap)
print("Gentoo:", avg_bill_depth_gentoo)

Adelie: 18.338494536780548
Chinstrap: 18.42058823529412
Gentoo: 14.999606206376153


# Insights
<hr>
 Chinstrap penguins have the deepest bills on average, which may relate to diet.

<h1>2 What is the average flipper length of Chinstrap penguins by sex?</h1>

In [23]:
# What is the average flipper length of Chinstrap penguins by sex?
chinstrap_df = penguins[penguins['species'] == 'Chinstrap']
chinstrap_df = chinstrap_df.dropna(subset=['sex', 'flipper_length_mm'])
chinstrap_df['sex'] = chinstrap_df['sex'].str.lower()

male_avg = chinstrap_df[chinstrap_df['sex'] == 'male']['flipper_length_mm'].mean()
female_avg = chinstrap_df[chinstrap_df['sex'] == 'female']['flipper_length_mm'].mean()

print("Male average flipper length:", male_avg)
print("Female average flipper length:", female_avg)

Male average flipper length: 199.91176470588235
Female average flipper length: 191.73529411764707


<h1>Insights</h1>
<hr>
Males have slightly longer flippers than females average of 199.9

<h1>3 In what year was the shortest flipper length recorded in an Adelie penguin?</h1>

In [24]:
# In what year was the shortest flipper length recorded in an Adelie penguin?
adelie_df = penguins[penguins['species'] == 'Adelie']
shortest_flipper_row = adelie_df[adelie_df['flipper_length_mm'] == adelie_df['flipper_length_mm'].min()]
shortest_year = shortest_flipper_row['year'].values[0]

print("Year:", shortest_year)

Year: 2007


<h1> Insights</h1>
<hr>
 An unusually short flipper was recorded in this year—possibly a young bird or measurement issue

<h1>Q4: What is the average body mass of penguins on Dream Island?</h1>

In [25]:

# Q4: What is the average body mass of penguins on Dream Island?
dream_penguins = penguins[penguins['island'] == 'Dream']

average_mass = dream_penguins['body_mass_g'].mean()

print(f"Dream Island - Average body mass: {average_mass:.2f} g")


Dream Island - Average body mass: 3712.90 g


<h1>Insights</h1>
<hr>

This tells us how heavy penguins usually are on Dream Island 
We can use this to compare penguin sizes across other islands or species

<h1>5. What is the average bill length of Gentoo penguins per year?</h1>

In [26]:
# Q5: Gentoo bill length per year
gentoo_2007 = penguins[(penguins['species'] == 'Gentoo') & (penguins['year'] == 2007)]['bill_length_mm'].mean()
gentoo_2008 = penguins[(penguins['species'] == 'Gentoo') & (penguins['year'] == 2008)]['bill_length_mm'].mean()
gentoo_2009 = penguins[(penguins['species'] == 'Gentoo') & (penguins['year'] == 2009)]['bill_length_mm'].mean()

print("2007:", gentoo_2007)
print("2008:", gentoo_2008)
print("2009:", gentoo_2009)




2007: 47.01470588235294
2008: 46.93695652173912
2009: 48.39595295055822


<h1>Insights</h1>
<hr>
The average bill length stayed fairly stable, showing no major shift over time.

<h1>6. How many Chinstrap penguins were recorded in 2008, and which one had the longest bill</h1>

In [27]:
# Q6: Total Chinstrap in 2008 and longest bill
chinstrap_2008 = penguins[(penguins['species'] == 'Chinstrap') & (penguins['year'] == 2008)]
total_count = chinstrap_2008.shape[0]
longest_bill = chinstrap_2008['bill_length_mm'].max()

print("Total Chinstrap in 2008:", total_count)
print("Longest bill:", longest_bill)



Total Chinstrap in 2008: 18
Longest bill: 54.2


<h1>Insights</h1>
<hr>
There was a healthy number of Chinstrap penguins in 2008. The longest bill might indicate an older

<h1>Q7: What is the most common penguin species found on Dream Island?</h1>

In [28]:
# Q7: What is the most common penguin species found on Dream Island?
dream_penguins = penguins[penguins['island'] == 'Dream']
species_counts = dream_penguins['species'].value_counts()
most_common_species = species_counts.idxmax()
count = species_counts.max()

print(f"The most common species on Dream Island is: {most_common_species} ({count} penguins)")

The most common species on Dream Island is: Chinstrap (68 penguins)


<h1>Insights</h1>
<hr>
 Knowing the most common species on an island helps us understand the habitat preferences of penguins.
In this case, it shows that Dream Island is a key habitat for the Chinstrap species.

# Q8: What is the smallest flipper length recorded on Biscoe Island?

In [29]:
# Q8: What is the smallest flipper length recorded on Biscoe Island?
smallest_flipper_biscoe = penguins[penguins['island'] == 'Biscoe']['flipper_length_mm'].min()
print("Smallest flipper length on Biscoe Island:", smallest_flipper_biscoe, "mm")



Smallest flipper length on Biscoe Island: 172.0 mm


# Insights
<hr>
Insight: This could belong to a young penguin still developing its swimming skills.

# Q9: What is the largest bill depth recorded on Torgersen Island?

In [30]:
# Q9: What is the largest bill depth recorded on Torgersen Island?
torgersen_df = penguins[penguins['island'] == 'Torgersen']
largest_bill_depth_torgersen = torgersen_df['bill_depth_mm'].max()

print("Largest bill depth on Torgersen Island:", largest_bill_depth_torgersen)

Largest bill depth on Torgersen Island: 21.5


# Insights
<hr>

 A deep bill can help crack tougher prey like crustaceans

# Q10: Which year had the highest average body mass of penguins on Dream Island?

In [31]:
# Q10: Which year had the highest average body mass of penguins on Dream Island?
dream_df = penguins[penguins['island'] == 'Dream']
year_mass_dream = dream_df.groupby('year')['body_mass_g'].mean()
year_max_dream = year_mass_dream.idxmax()

print("Year with highest average body mass on Dream Island:", year_max_dream)

Year with highest average body mass on Dream Island: 2008


<h1>Insights</h1>
<hr>

This year might have had better food supply or weather conditions

# Q11: How many Chinstrap penguins are found on Dream Island?

In [32]:
# Q11: How many Chinstrap penguins are found on Dream Island?
count_chinstrap_dream = penguins[(penguins['island'] == 'Dream') & (penguins['species'] == 'Chinstrap')]['species'].count()
print("Number of Chinstrap penguins on Dream Island:", count_chinstrap_dream)

Number of Chinstrap penguins on Dream Island: 68


<h1>Insights</h1>
<hr>

Chinstrap penguins are well-adapted to Dream Island and are common there

# Q12: What is the average body mass of male penguins on Biscoe Island?

In [33]:
# Q12: What is the average body mass of male penguins on Biscoe Island?
biscoe_df = penguins[penguins['island'] == 'Biscoe']
biscoe_males = biscoe_df[biscoe_df['sex'] == 'male']
biscoe_females = biscoe_df[biscoe_df['sex'] == 'female']
avg_mass_biscoe_males = biscoe_males['body_mass_g'].mean()
avg_mass_biscoe_females = biscoe_females['body_mass_g'].mean()
print(f"Average body mass of male penguins on Biscoe Island: {avg_mass_biscoe_males:.2f} g")
print(f"Average body mass of female penguins on Biscoe Island: {avg_mass_biscoe_females:.2f} g")

Average body mass of male penguins on Biscoe Island: 5070.76 g
Average body mass of female penguins on Biscoe Island: 4319.38 g


<h1>Insights</h1>
<hr>

 Male penguins on Biscoe Island are generally heavier than females

# Q13: What is the longest bill length recorded on Dream Island?

In [34]:
# Q13: What is the longest bill length recorded on Dream Island?
longest_bill_dream = dream_df['bill_length_mm'].max()
print("Longest bill length on Dream Island:", longest_bill_dream)

Longest bill length on Dream Island: 58.0


# Insights
<hr>

A long bill helps penguins catch more food efficiently on Dream Island